tsdb/engine/tsm1/compact_test.go

```diff
@@ -2306,66 +2323,196 @@ func TestDefaultPlanner_PlanOptimize_Level4(t *testing.T) {
+		expFiles = append(expFiles, file)
+	}
+
+	_, cgLen := cp.PlanLevel(1)
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(1)")
+	_, cgLen = cp.PlanLevel(2)
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(2)")
+	_, cgLen = cp.PlanLevel(3)
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(3)")
+
+	tsmP, pLenP := cp.Plan(time.Now().Add(-time.Second))
+	require.Zero(t, len(tsmP), "compaction group; Plan()")
+	require.Zero(t, pLenP, "compaction group length; Plan()")
+
+	tsm, pLen, gLen := cp.PlanOptimize()
+	require.Equal(t, 1, len(tsm), "compaction group")
+	require.Equal(t, int64(len(tsm)), pLen, "compaction group length")
+	require.Equal(t, int64(1), gLen, "generation of TSM files")
+	require.Equal(t, len(expFiles), len(tsm[0]), "expected TSM files")
+}
+
+// This test is added to account for a single generation that has a group size
+// under 2 GB and has less then level 4 files it should be further compacted to a single file.
+func TestDefaultPlanner_PlanOptimize_SmallSingleGenerationUnderLevel4(t *testing.T) {
+	// ~650 MB total group size
+	data := []tsm1.FileStat{
 		{
-			Path: "05-03.tsm1",
-			Size: 2 * 1024 * 1024 * 1024,
+			Path: "01-02.tsm1",
+			Size: 300 * 1024 * 1024,
 		},
 		{
-			Path: "06-03.tsm1",
-			Size: 2 * 1024 * 1024 * 1024,
+			Path: "01-03.tsm1",
+			Size: 200 * 1024 * 1024,
 		},
 		{
-			Path: "07-04.tsm1",
-			Size: 2 * 1024 * 1024 * 1024,
+			Path: "01-04.tsm1",
+			Size: 100 * 1024 * 1024,
 		},
+	}
+
+	cp := tsm1.NewDefaultPlanner(
+		&fakeFileStore{
+			PathsFn: func() []tsm1.FileStat {
+				return data
+			},
+		}, tsdb.DefaultCompactFullWriteColdDuration,
+	)
+
+	expFiles := make([]tsm1.FileStat, 0)
+	for _, file := range data {
+		expFiles = append(expFiles, file)
+	}
+
+	_, cgLen := cp.PlanLevel(1)
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(1)")
+	_, cgLen = cp.PlanLevel(2)
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(2)")
+	_, cgLen = cp.PlanLevel(3)
+	require.Zero(t, cgLen, "compaction group length; PlanLevel(3)")
+
+	tsmP, pLenP := cp.Plan(time.Now().Add(-time.Second))
+	require.Zero(t, len(tsmP), "compaction group; Plan()")
+	require.Zero(t, pLenP, "compaction group length; Plan()")
+
+	tsm, pLen, gLen := cp.PlanOptimize()
+	require.Equal(t, 1, len(tsm), "compaction group")
+	require.Equal(t, int64(len(tsm)), pLen, "compaction group length")
+	require.Equal(t, int64(1), gLen, "generation of TSM files")
+	require.Equal(t, len(expFiles), len(tsm[0]), "expected TSM files")
+}
+
+// This test is added to account for a single generation that has a group size
+// under 2 GB and all files at max default points per block of 1000.
+// This should be planned for compaction at a more aggressive points per block.
+func TestDefaultPlanner_FullyCompacted_SmallSingleGeneration(t *testing.T) {
+	// ~650 MB total group size
+	data := []tsm1.FileStat{
 		{
-			Path: "08-04.tsm1",
-			Size: 2 * 1024 * 1024 * 1024,
+			Path: "01-05.tsm1",
```

Comment:
As recommended above, order these in reverse, as well, to ensure file size order is unimportant.